### PR TITLE
refactor: initial full cleanup

### LINKED ISSUE #150: Account for All Edge Cases
We have run into multiple failures on CMUEats because the Dining API doesn't account for all edge cases. I have provided a detailed outline of edge cases to account for in https://github.com/ScottyLabs/dining-api/wiki/Timing-Cases.

To account for these, we need to fix the timing logic in https://github.com/ScottyLabs/dining-api/blob/main/src/parser/diningParser.ts, starting at the following line:

```const timeBuilder = new TimeBuilder();```

### PR DESCRIPTION
**Pull request by cirex-web**
# Description

No change in functionality, just a lot of moving code around for a hopefully more readable/debuggable codebase.

Conveniently closes #150 (tests aren't in this PR but I ran them here to make sure)
image

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [ ] Test A
- [ ] Test B

**Test Configuration**:

- Node.js version:
- Python version:
- Desktop/Mobile:
- OS:
- Browser:

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by GhostOf0days** [comment]
Heads up while still in draft.

Some logic in dealing with schedule strings was moved to various locations. Merge interval logic has caused failures on edge cases before. Therefore, please use a version of this block of logic (its complexity is needed for edge cases) rather than moving its logic or keep most of it in one area in this PR, so easier to review and don't accidentally break API.  Can move in new PR after.

More specifically, this logic is very important.
```
      const timeInfoType = determineTimeInfoType(timeStr);

      if (
        timeInfoType === TimeInfoType.CLOSED ||
        timeInfoType === TimeInfoType.TWENTYFOURHOURS
      ) {
        const scheduleString = `${dayStr.trim()}, ${timeStr}`;
        addedSchedules.add(scheduleString);
        timeBuilder.addSchedule([dayStr.trim(), dateStr.trim(), timeStr]);
      } else if (timeInfoType === TimeInfoType.TIME) {
        const timeSlots = timeStr.split(/[,;]/).map((slot) => slot.trim());

        // Sort time slots based on opening time
        timeSlots.sort((a, b) => {
          const [aStart, aEnd] = a.split("-").map((time) => time.trim());
          const [bStart, bEnd] = b.split("-").map((time) => time.trim());
          const startComparison = aStart.localeCompare(bStart);
          if (startComparison !== 0) {
            return startComparison;
          }
          return bEnd.localeCompare(aEnd); // Reverse order for end times
        });

        // Merge overlapping, contained, and duplicate time slots
        const mergedTimeSlots = [];
        let prevSlot = null;
        for (const timeSlot of timeSlots) {
          const [start, end] = timeSlot.split("-").map((time) => time.trim());

          if (prevSlot && start === prevSlot.start) {
            // If the current time slot has the same opening time as the previous one
            // Update the previous slot with the later closing time
            if (end > prevSlot.end) {
              prevSlot.end = end;
            }
          } else {
            mergedTimeSlots.push({ start, end });
            prevSlot = { start, end };
          }
        }

        // Format and add merged time slots
        mergedTimeSlots.forEach((slot) => {
          let { start, end } = slot;

          // Handle case where end time is 12:00 AM
          if (/12:00 AM$/i.test(end)) {
            end = end.replace(/12:00 AM$/i, "11:59 PM");
          }

          const scheduleString = `${dayStr.trim()}, ${start} - ${end}`;
          if (!addedSchedules.has(scheduleString)) {
            addedSchedules.add(scheduleString);
            timeBuilder.addSchedule([
              dayStr.trim(),
              dateStr.trim(),
              `${start} - ${end}`,
            ]);
          }
        });
      }
    }

    builder.setTimes(timeBuilder.build());
    ```

**Comment by cirex-web** [comment]
Thanks for pointing that out! I believe that the above functionality has all been shifted to the following two code snippets:
```ts
export function sortAndMergeTimeRanges(timeRanges: ITimeRange[]) {
  timeRanges.sort((range1, range2) =>
    compareTimeMoments(range1.start, range2.start)
  );
  const mergedRanges: ITimeRange[] = [];

  for (const timeRange of timeRanges) {
    const lastTimeRange = mergedRanges.length
      ? mergedRanges[mergedRanges.length - 1]
      : undefined;
    if (lastTimeRange && lastTimeRange.end >= timeRange.start) {
      if (compareTimeMoments(lastTimeRange.end, timeRange.end) > 0) {
        lastTimeRange.end = timeRange.end; // join current range with last range
        continue;
      }
    }

    mergedRanges.push(timeRange);
  }
  return mergedRanges;
}
```
and
```ts
function getTimeRangesFromTimeRow(time: ITimeRowAttributes) {
  if (time.day === undefined) {
    throw new Error("Cannot convert when day is not set");
  }
  const allRanges: ITimeRange[] = [];
  for (const range of time.times ?? []) {
    if (range.end.hour === 0 && range.end.minute === 0) {
      range.end.hour = 23;
      range.end.minute = 59; // roll back from 12AM to 11:59 previous day <- specifically this
    }

    const spillToNextDay =
      range.start.hour * 60 + range.start.minute >
      range.end.hour * 60 + range.end.minute;

    allRanges.push({
      start: {
        day: time.day,
        hour: range.start.hour,
        minute: range.start.minute,
      },
      end: {
        day: spillToNextDay ? getNextDay(time.day) : time.day,
        hour: range.end.hour,
        minute: range.end.minute,
      },
    });
  }
  return allRanges;
}
```
so rather than filtering by type and sorting before time parsing, we just do it afterwards with `ITimeRange`s

I can also add unit tests for all of the corner cases you provided in an earlier Issue

**Comment by cirex-web** [comment]
Thanks for pointing that out! I believe that the above functionality has all been shifted to the following two code snippets:
```ts
export function sortAndMergeTimeRanges(timeRanges: ITimeRange[]) {
  timeRanges.sort((range1, range2) =>
    compareTimeMoments(range1.start, range2.start)
  );
  const mergedRanges: ITimeRange[] = [];

  for (const timeRange of timeRanges) {
    const lastTimeRange = mergedRanges.length
      ? mergedRanges[mergedRanges.length - 1]
      : undefined;
    if (
      lastTimeRange &&
      compareTimeMoments(lastTimeRange.end, timeRange.start) >= 0
    ) {
      if (compareTimeMoments(timeRange.end, lastTimeRange.end) > 0) {
        lastTimeRange.end = timeRange.end; // join current range with last range
      }
    } else {
      mergedRanges.push(timeRange);
    }
  }
  console.log(timeRanges, mergedRanges);
  return mergedRanges;
}
```
and
```ts
function getTimeRangesFromTimeRow(time: ITimeRowAttributes) {
  if (time.day === undefined) {
    throw new Error("Cannot convert when day is not set");
  }
  const allRanges: ITimeRange[] = [];
  for (const range of time.times ?? []) {
    if (range.end.hour === 0 && range.end.minute === 0) {
      range.end.hour = 23;
      range.end.minute = 59; // roll back from 12AM to 11:59 previous day <- specifically this
    }

    const spillToNextDay =
      range.start.hour * 60 + range.start.minute >
      range.end.hour * 60 + range.end.minute;

    allRanges.push({
      start: {
        day: time.day,
        hour: range.start.hour,
        minute: range.start.minute,
      },
      end: {
        day: spillToNextDay ? getNextDay(time.day) : time.day,
        hour: range.end.hour,
        minute: range.end.minute,
      },
    });
  }
  return allRanges;
}
```
so rather than filtering by type and sorting before time parsing, we just do it afterwards with `ITimeRange`s

I can also add unit tests for all of the corner cases you provided in an earlier Issue

File: src/utils/timeUtils.ts [hunk 1]
```diff
@@ -47,160 +13,43 @@ export function getNextDay(day: DayOfTheWeek): DayOfTheWeek {
...
-  }
-  if (
-    Array.isArray(
-      input.match(/\d\d?:\d\d\s?(?:am|pm)\s?-\s?\d\d?:\d\d\s?(?:am|pm)/)
-    )
-  ) {
-    return TimeInfoType.TIME;
-  }
-  throw new Error("Could not determine time info type");
+  console.log(timeRanges, mergedRanges);
```
Comment (GhostOf0days) [inline src/utils/timeUtils.ts:53]: No issues that I see. Thanks for your PR. However, one thing I wanted to make sure, is if this print statement was left in on accident or is it left for later debugging on the time parsing algo if an issue pops up.
